record/program/src/error.rs

```diff
@@ -0,0 +1,27 @@
+//! Error types
+
+use num_derive::FromPrimitive;
+use solana_program::{decode_error::DecodeError, program_error::ProgramError};
+use thiserror::Error;
+
+/// Errors that may be returned by the program.
+#[derive(Clone, Debug, Eq, Error, FromPrimitive, PartialEq)]
+pub enum RecordError {
+    /// Incorrect owner provided on update or delete
+    #[error("Incorrect owner provided on update or delete")]
+    IncorrectOwner,
+
+    /// Calculation overflow
+    #[error("Calculation overflow")]
+    Overflow,
```

Comment:
I'd prefer to make a new error for `CalculationError` since `UnbalancedInstruction` isn't really the same concept.  I'm happy to do everything to remove `RecordError` though!  Also, not all `InstructionError`s are mapped to `ProgramError`, so we'd have to add those mappings -- is there a reason for that?

Reply:
A new `ProgramError::Overflow` sounds good now that you mention it.

I'm not sure why we don't have all `InstructionErrors`s mapped into `ProgramError`.  Might have just _happened_ unintentionally.  Do you have any thoughts here Jack?   Mapping most/all of them into `ProgramError` seems nice.  It looks like we might need to be careful with the rollout of such a change though, as it seems to me like we could risk a bank hash mismatch if a program generates a new `ProgramError` before the supermajority has upgraded to know about it

Reply:
tagging @jackcmay in case he didn't see ^

Reply:
Some instruction errors don't make sense coming from programs but I don't see any reason not to map them anyway.  @mvines What bankhash issue are you worried about?  Something more than adding a new instruction error?

Reply:
Unknown program errors are mapped to/from `ProgramError::Custom` [here](https://github.com/solana-labs/solana/blob/5d31763a3b48a9199861fad09166a0d61719fd52/sdk/program/src/program_error.rs#L157).  If we add a new enum entry to ProgramError, the nodes that know about that entry will no longer map it to/from `ProgramError::Custom`.   If a program generates such an error before 66% of the cluster has upgraded to a software version that knows about it, I'm worried we'll get a bank hash mismatch (since the transaction result is encoded in the bank hash) and possibly kill the cluster